vnext/Microsoft.ReactNative/Views/TextInputViewManager.cpp

```diff
@@ -712,9 +712,40 @@ void TextInputShadowNode::SetText(const winrt::Microsoft::ReactNative::JSValue &
         auto oldValue = textBox.Text();
         auto newValue = asHstring(text);
         if (oldValue != newValue) {
-          textBox.Text(newValue);
+          if (!textBox.IsReadOnly()) {
+            uint32_t diffStartIndex = 0;
+            uint32_t diffEndIndex = 0;
+            // Replacing the entire string with the new value resets the text history.
+            // Generally speaking, this is undesirable. To mitigate this issue, we can
+            // replace text using the XAML TextBox::SelectedText property, which retains
+            // undo / redo history. To do so, we use a simple algorithm that finds a
+            // single diff in the values:
+            // 1. Find first character that mismatches iterating forwards
+            while (diffStartIndex < oldValue.size() && diffStartIndex < newValue.size() &&
+                   oldValue[diffStartIndex] == newValue[diffStartIndex])
+              diffStartIndex++;
+            // 2. Find last character the mismatches beyond the first mismatch iterating backwards
+            while (diffEndIndex < oldValue.size() - diffStartIndex && diffEndIndex < newValue.size() - diffStartIndex &&
+                   oldValue[oldValue.size() - diffEndIndex - 1] == newValue[newValue.size() - diffEndIndex - 1])
+              diffEndIndex++;
```

Comment:
Would be good to have test cases for this to ensure that this produces the right results for different string values. Seems like something unit testable on its own (no need for UI automation), and also would benefit from UI automation. There might not be the right infrastructure for that now, but if it were that'd be the right call here.

Also feels like this block should be a named function/lambda for clarity ("findUnchangedRange" or whatever)